pkg/controller/externalippool/validate.go

```diff
@@ -130,115 +120,41 @@ func validateIPRangesAndSubnetInfo(externalIPPool crdv1beta1.ExternalIPPool, exi
 
 			if ipRange.CIDR != "" {
 				key = fmt.Sprintf("range [%s] of pool %s", ipRange.CIDR, pool.Name)
-				cidr, _ := parseIPRangeCIDR(ipRange.CIDR)
+				cidr, _ := validation.ParseIPRangeCIDR(ipRange.CIDR)
 				start, end = utilip.GetStartAndEndOfPrefix(cidr)
-
 			} else {
 				key = fmt.Sprintf("range [%s-%s] of pool %s", ipRange.Start, ipRange.End, pool.Name)
-				start, end, _ = parseIPRangeStartEnd(ipRange.Start, ipRange.End)
-
+				start, end, _ = validation.ParseIPRangeStartEnd(ipRange.Start, ipRange.End)
 			}
-			combinedRanges[key] = [2]netip.Addr{start, end}
+			existingRanges[key] = [2]netip.Addr{start, end}
 		}
 	}
 
 	for _, ipRange := range ipRanges {
 		var key string
 		var start, end netip.Addr
+		var errMsg string
 
 		if ipRange.CIDR != "" {
 			key = fmt.Sprintf("range [%s]", ipRange.CIDR)
-			cidr, errMsg := parseIPRangeCIDR(ipRange.CIDR)
+			cidr, errMsg := validation.ParseIPRangeCIDR(ipRange.CIDR)
```

Comment:
I feel like we do end up with quite a bit of code duplication now (with the shared `ValidateIPRangesAndSubnetInfo`). Maybe this can be reduced a bit, as we don't need to handle the error case here again?

Another idea would be to have a separate function to normalize `Spec.IPRanges`, so we only need to do it once. Maybe a separate helper type would help with readability:
```go
type NormalizedIPRange struct {
    Start netip.Addr
    End netip.Addr
    Origin string  // describes where the range comes from (corresponds to the key of the `existingRanges` map)
}
```

Now when you check for overlap, you can just use `NormalizedIPRange` consistently

Reply:
Updated. Please take a look.